Performance counters for NVIDIA hardware

reinsteam opened the pull request:
# Description

Added an option to query NVIDIA performance counters using *NVIDIA Performance Monitor API* distributed as a part of [**NVIDIA PerfKit**](https://developer.nvidia.com/nvidia-perfkit) (version 4.4.0 was used for testing).

![](https://pbs.twimg.com/media/DYk4854VwAAXgRH.jpg:large)

## Limitation
- Currently, getting data from counters works only with D3D11 replay.
- *NvPmApi* could be used with D3D10, OpenGL, D3D9 APIs.

## Potential Improvements
- Organizing counters in more finer-grained groups than default ones coming from API side.
- Unfortunately, NvPmApi also returns outdated/deprecated/non-relevant counters in a list of available ones. Creation of GPU generation lists of valid counters could be a useful improvement.
- Integration with [**PerfWorks**](https://developer.nvidia.com/perfworks) would be an interesting addition, but it is not available publicly.

## Additional Requirements for Packaging
- Before distributing this feature to the end user, it is necessary to include `NvPmApi.Core.dll` from [**NVIDIA PerfKit**](https://developer.nvidia.com/nvidia-perfkit) into the final package.
- Currently, `NvPmApi.Core.dll` is supposed to be placed in a folder with *RenderDoc executable*.
- Probably, the more reasonable place would be `./plugins/nv/counters` (require a change of the loading path in `nv_counters.cpp`)

## Pull Request Notes
- Header files for *NvPmApi*, its release notes and copyright notices are included in PR.
- Build verification [was done](https://ci.appveyor.com/project/reinsteam/renderdoc) in a local branch.

reinsteam committed:
Added a project with NVidia counters and NvPmApi header from NV Perfkit

- nv_counters.h/.cpp load NVidia counters using NvPmApi 4.4.0

reinsteam committed:
Plugged in NV counters into D3D11Replay.

- Counters are displayed in RenderDoc 'Performance counter selection' window

reinsteam committed:
Implemented data retrieval from selected NV counters

baldurk commented on renderdoc/driver/d3d11/d3d11_counters.cpp:
```diff
@@ -332,6 +348,73 @@ void D3D11Replay::FillTimersAMD(uint32_t &eventStartID, uint32_t &sampleIndex,
   }
 }
 
+static void FlushGPU(WrappedID3D11Device *pDevice, WrappedID3D11DeviceContext *pImmediateContext)
```
Can this function be deleted and instead have code use the `D3D11Replay::SerializeImmediateContext` a bit above instead?

baldurk commented on renderdoc/driver/d3d11/d3d11_counters.cpp:
```diff
@@ -332,6 +348,73 @@ void D3D11Replay::FillTimersAMD(uint32_t &eventStartID, uint32_t &sampleIndex,
+
+    FlushGPU(m_pDevice, m_pImmediateContext);
+
+    m_pNVCounters->EndSample(sampleIndex);
```
It seems strange to flush/serialize the GPU *before* ending the sample. Is this definitely what you want to do? Instead I'd expect to either flush only before starting each sample, or else do `flush -> begin -> replaylog -> end -> flush`.

reinsteam replied:
The flush *before* `EndSample` is the necessary thing according to *NVIDIA PerfKit User Guide 4.2.0* (see page 5, section *Experiment Mode*). The flush *before* `BeginSample` is just to make sure that we serialized everything right before draw/dispatch call. Also, similar approach could be found in a sample code (`PerfKit/samples/NvPmSampleDX`) for the *PerfKit*.

baldurk replied:
Fair enough! It seems strange to me so I just wanted to check it was on purpose.

baldurk commented on renderdoc/driver/d3d11/d3d11_replay.cpp:
```diff
@@ -156,20 +157,30 @@ void D3D11Replay::CreateResources()
 
-  AMDCounters *counters = new AMDCounters();
+  AMDCounters *pAMDCounters = new AMDCounters();
+  NVCounters *pNVCounters = new NVCounters();
```
I know I'm kind of inconsistent about this in a lot of older written code with `m_pFoo` members but I generally don't use hungarian pointer prefix for locals like this, so `AMDCounters` and `NVCounters` would be preferred.

baldurk commented on renderdoc/driver/ihv/nv/nv_counters.cpp:
```diff
@@ -0,0 +1,380 @@
+#include "nv_counters.h"
```
The two new files need the standard MIT License copyright comment at the top, with just `Copyright (c) 2018 Baldur Karlsson`

baldurk commented on renderdoc/driver/ihv/nv/nv_counters.cpp:
```diff
@@ -0,0 +1,380 @@
+
+int NvPmCountCounters(NVPMCounterID unCounterID, const char *pcCounterName, void *pUserData)
+{
+  uint32_t *pNumCounters = static_cast<uint32_t *>(pUserData);
```
Please prefer normal casts `= (uint32_t *)pUserData;` over C++ style casts. There are a few other instances in this file.

baldurk commented on renderdoc/driver/ihv/nv/nv_counters.cpp:
```diff
@@ -0,0 +1,380 @@
+
+  char Description[512];
+  NVPMUINT DescriptionSize = sizeof(Description);
+  pEnumCtx->mNvPmApi->GetCounterDescription(unCounterID, Description, &DescriptionSize);
```
Is it possible to call this function twice, to first get the size and then get the actual text? In case the description is longer than 512 characters.

baldurk commented on renderdoc/driver/ihv/nv/nv_counters.h:
```diff
@@ -0,0 +1,51 @@
+private:
+  bool Init(void);
+
+  void *mNvPmLib;
```
For members either use `m_` prefix or not at all, the plain `m` prefix is not used anywhere else.

baldurk commented on renderdoc/driver/ihv/nv/nv_counters.cpp:
```diff
@@ -0,0 +1,380 @@
+  std::for_each(counters.begin(), counters.end(), [
+                                                      &selExternalIds = mSelectedExternalIds,
+                                                      &selInternalIds = mSelectedInternalIds,
+                                                      &internalIds = mInternalIds
```
Is there any reason you're using `std::for_each` and a callback here instead of just a simple for loop? You can use `for(GPUCounter counter : counters)` if you prefer that style here.

baldurk commented on renderdoc/driver/ihv/nv/nv_counters.cpp:
```diff
@@ -0,0 +1,380 @@
+          NVPMRESULT result =
+              mNvPmApi->GetCounterValueUint64(mNvPmCtx, internalId, i, &Value, &Cycles, &Overflow);
+
+          double Ratio = static_cast<double>(Value) / static_cast<double>(Cycles);
```
I don't have the values in front of me so I don't really know, but is it possible it would be better to report these as percentages instead of a unit-less ratio? Since "0.6" is maybe less obvious than "60%" if the value is e.g. what proportion of the time was spent doing X. Since it's always a ratio compared to cycles this feels like maybe a better fit, but you can tell better than me.

reinsteam replied:
Unfortunately, information about units for counters is almost non-existing thing. `NvPmApi` supports `float64` / `uint64` counters with display types `raw` / `ratio`. `ratio`-counters are usually defined by two `uint64` variables and it's usually something like `metric/frequency`. When the actual value for `ratio`-counter is computed, it can be a ratio (instructions per shader invocation, for example), a percentage (l2 hit rate) or a normalized ratio. The same thing is true for `raw` `float64` counters. Real meaning could be understood by a user from a description string.

baldurk replied:
Yeh that makes sense, I hoped that maybe it would be somewhat consistent but I understand the problem.

baldurk reviewed (changes requested):
Thanks for this! I've made some notes on the code about possible improvements and some code-style things but hopefully nothing too horrible. Also could you please rebase the clang-formatting changes into the relevant earlier changes rather than having them as a separate commit?

A couple of other higher level things:

> Unfortunately, NvPmApi also returns outdated/deprecated/non-relevant counters in a list of available ones. Creation of GPU generation lists of valid counters could be a useful improvement.

Can you go into more detail about this? For me or anyone else who might want to do this in future. I don't understand exactly what you mean - you're saying that you'd somehow generate the list of counters on the GPU itself instead of the CPU?

> Before distributing this feature to the end user, it is necessary to include NvPmApi.Core.dll from NVIDIA PerfKit into the final package.

I'll need to check the licensing to make sure this is OK, but I can make sure this gets into builds from now on. However...

> Probably, the more reasonable place would be ./plugins/nv/counters (require a change of the loading path in nv_counters.cpp)

I'd like you to make this change for this PR, please, since then it will work with if I drop the plugin in the builds. It should be pretty easy since if you look at `amd_counters.cpp` you just need to call `LocatePluginFile` to do the search path logic. Usually I then fall-back to a plain `LoadModule` in case the plugin isn't found otherwise, so it would be mostly the same.

reinsteam commented:
Thank you for taking the time reviewing this change! Do you prefer to keep changes for the PR as a separate commit or have them squashed?

> > Unfortunately, NvPmApi also returns outdated/deprecated/non-relevant counters in a list of available ones. Creation of GPU generation lists of valid counters could be a useful improvement.
>
> Can you go into more detail about this? For me or anyone else who might want to do this in future. I don't understand exactly what you mean - you're saying that you'd somehow generate the list of counters on the GPU itself instead of the CPU?

Sorry for the confusion, my bad, haven't noticed typos here. Initially, it should be `CPU generated lists of valid counters`. The idea was that it's possible to create, for example, lists of counters on CPU side per GPU generation (Fermi/Kepler/Maxwell) or even by purpose (l1-cache traffic / l2-cache traffic) to improve grouping of available counters and remove outdated ones.

> I'll need to check the licensing to make sure this is OK, but I can make sure this gets into builds from now on. However...
>> Probably, the more reasonable place would be ./plugins/nv/counters (require a change of the loading path in nv_counters.cpp)
>
>I'd like you to make this change for this PR, please, since then it will work with if I drop the plugin in the builds. It should be pretty easy since if you look at amd_counters.cpp you just need to call LocatePluginFile to do the search path logic. Usually I then fall-back to a plain LoadModule in case the plugin isn't found otherwise, so it would be mostly the same.

I will update the PR with those changes, just wanted to make sure that you are fine with that location.

Comments regarding `FlushGPU` and units for displaying counters will be in place.

baldurk commented on renderdoc/driver/ihv/nv/nv_counters.h:
```diff
@@ -0,0 +1,75 @@
+  CounterDescription GetCounterDescription(GPUCounter counterID) const
+  {
+    const uint32_t LocalId =
+        static_cast<uint32_t>(counterID) - static_cast<uint32_t>(GPUCounter::FirstNvidia);
```
A couple of `static_cast` still left here.

reinsteam replied:
Missed that one. Thanks! Pushed an update

baldurk commented:
LGTM now, one minor tweak left over but otherwise this is great, thanks!

reinsteam committed:
Various fixes for PR

- Various style-related fixes
- Loading NvPmApi.dll from plugins directories
- Added copyright notice to newly added files
- Replaced local fixed size array of `char` with std::vector<char>
- Removed `FlushGPU`

GraphicsCoder commented:
Hi folks,

There're some issues related NV counters by using the latest version of RenderDoc. Do I miss something?(I've added NvPmApi.Core.dll in the executable table. But there're some waning message popped up as below(The result is also incorrect):
![nv_counter_issue](https://user-images.githubusercontent.com/13574689/40608379-def6796c-629d-11e8-80ba-20eea6bd477c.PNG)
Note: can you help to attach the DLL, which we may use the different version?
Thanks.

baldurk commented:
This seems like an internal nvapi error. I don't think I can help diagnose this, it would either be nvidia or maybe @reinsteam has an idea.

Currently renderdoc builds don't distribute nv perfkit because I haven't clarified whether the license allows that or not.

GraphicsCoder commented:
Ok, thanks for your feedback so quickly. It seems like an internal nvapi error.(I used 980Ti to reproduce this issue.)
So, @reinsteam please take a look if you are available. Thanks.
